File: Algorithms on Graphs/week2_decomposition2/2_toposort/toposort.py
# Uses python3

import sys


def dfs(adj, used, order, x):
    # write your code here
    used[x] = 1

    for j in adj[x]:
        if used[j] == 0:
            order = dfs(adj, used, order, j)

    order.append(x)
    return order


def toposort(adj):
    # print(adj)
    used = [0] * len(adj)
    order = []
    # write your code here
    for i in range(len(adj)):
        if used[i] == 0:
            order = dfs(adj, used, order, i)

    return reversed(order)


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    order = toposort(adj)
    for x in order:
        print(x + 1, end=' ')

File: Algorithms on Graphs/week2_decomposition2/2_toposort/Toposort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Scanner;

public class Toposort {
    private static ArrayList<Integer> toposort(ArrayList<Integer>[] adj) {

        int used[] = new int[adj.length];
        ArrayList<Integer> order = new ArrayList<Integer>();        // arraylist for storing order of node visit
        Arrays.fill(used, 0);

        for(int i=0; i < adj.length; i++)                           // DFS
        {
            if(used[i] == 0)
            {
                dfs(adj, used, order, i);
            }
        }
        return order;                                               // return order
    }

    private static void dfs(ArrayList<Integer>[] adj, int[] used, ArrayList<Integer> order, int s) {
      //write your code here
        used[s] = 1;                            // make node as visited

        for (int i: adj[s])                    // iterate through adjacent nodes
        {
            if(used[i] == 0)
            {
                dfs(adj, used, order, i);      // if not visited then explore
            }
        }
        order.add(0, s);                 // add element in order to keep track of order of visiting of node in TopoSort
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        ArrayList<Integer> order = toposort(adj);
        for (int x : order) {
            System.out.print((x + 1) + " ");
        }
    }
}

